alphas/alpha_functions.py

```diff
@@ -29,3 +31,198 @@ def alpha_001(data: pd.DataFrame, ticker: str) -> pd.DataFrame:
+    in the minimum of three values into tsargmin(x,3), rather than tsargmin(x,9).
+    """
+
+    def bscore(data: pd.Dataframe, b: int):
+        mean = data["Adj Close"].iloc[-b:].mean()
+        std = data["Adj Close"].iloc[-b:].std()
+        data = data.dropna()
+        bollinger_score = ((data["Adj Close"][-b:] - mean) / std).dropna()
+        return bollinger_score
+
+    def ts_argmin(x, d):
+        if isinstance(x, np.ndarray):
+            x = pd.Series(x)
+        elif not isinstance(x, pd.Series):
+            raise TypeError(type(x))
+        return x.rolling(d).apply(lambda arr: arr.argmax(), raw=True)
+
+    def ts_argmin_2(x, d):
+        """augmented tsargmin function to find the timeseries's minimum index looking
+        back at 3 values. So tsargmin(x,3) rathern than tsargmin(x,9)"""
+        if isinstance(x, np.ndarray):
+            x = pd.Series(x)
+        elif not isinstance(x, pd.Series):
+            raise TypeError(type(x))
+        return 1 + np.argmin(x[-d:])
+
+    def ts_max(x, d):
+        if isinstance(x, np.ndarray):
+            x = pd.Series(x)
+        elif not isinstance(x, pd.Series):
+            raise TypeError(type(x))
+        return x.rolling(d).max()
+
+    def tsrank_a(data: pd.DataFrame, a: int) -> pd.DataFrame:
+        try:
+            return stats.rankdata(data[-a:], method="average", nan_policy="omit")[-1]
+        except:
+            raise Exception(f"Data is not long enough for lookback period {a}")
+
+    def plus(data_a: pd.DataFrame, data_b: pd.DataFrame) -> pd.DataFrame:
+        try:
+            return data_a + data_b
+        except:
+            raise Exception("The summands are not the same datatypes and cannot be summed")
+
+    close = data["Adj Close"]
+    bollinger_score = bscore(data, 24)
+    tsargmin_1 = ts_argmin(bollinger_score, 21)
+    tsargmin_2 = ts_argmin_2(tsargmin_1, 3)
+    print(tsargmin_2)
+    tsmax = (ts_max(close, 6)).dropna()
+    inv_rank = tsrank_a(tsmax, 5) * -1
+    raw_signal = plus(inv_rank, tsargmin_2)
+    return raw_signal
+
+
+def alpha_015(data: pd.DataFrame, ticker=str):
+    def minus(data_a: pd.DataFrame, data_b: pd.DataFrame) -> pd.DataFrame:
+        """Difference of two dataframes"""
+        try:
+            return data_a - data_b
+        except:
+            raise Exception("The summands are not the same datatypes and cannot be subtracted")
+
+    def div(data_a: pd.Series, data_b: pd.Series) -> pd.DataFrame:
+        """Quotient of two columns"""
+        try:
+            return data_a / data_b
+        except:
+            raise Exception("The summands are not the same datatypes and cannot be subtracted")
+
+    close = data["Close"].iloc[-1]
+    open = data["Open"].iloc[-1]
+    high = data["High"].iloc[-1]
+    low = data["Open"].iloc[-1]
+    difference_1 = open - close
+    difference_2 = high - low
+    raw_signal = div(difference_1 / difference_2)
```

Comment:
you also don't need the extra variable initializations here. Could you do the same for the other alphas?